contracts/epochs/FlowEpoch.cdc

```diff
@@ -432,6 +559,97 @@ access(all) contract FlowEpoch {
+            collectorClusters: [[String]],
+            clusterQCVoteData: [FlowClusterQC.ClusterQCVoteData],
+            dkgPubKeys: [String],
+            nodeIDs: [String])
+        {
+            pre {
+                FlowEpoch.isValidPhaseConfiguration(stakingEndView-startView+1, FlowEpoch.configurableMetadata.numViewsInDKGPhase, endView-startView+1):
+                    "Invalid startView, stakingEndView, and endView configuration"
+            }
+ 
+            /// sanity check all nodes should have a weight > 0
+            for nodeID in nodeIDs {
+                assert(
+                    FlowIDTableStaking.NodeInfo(nodeID: nodeID).initialWeight > 0, 
+                    message: "all nodes in node ids list for recovery epoch must have a weight > 0"
+                )
+            }
+
+            // When we are in EFM, the Epoch Smart Contract's state has likely diverged from the Protocol State, 
+            // which orchestrates the behaviour of the nodes.  Any staking, or Epoch Setup (incl. DKG) that the Epoch smart
+            // contract is attempting to run are probably based on obsolete data. Therefore, we just abort them.
+            // With the recovery epoch, we effectively overwrite the state of the Epoch Smart Contract, aligning it with
+            // the Protocol State that the network currently operates with. 
+            if FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION {
+                // abort staking auction 
+                FlowEpoch.borrowStakingAdmin().endStakingAuction()
+            } else {
+                /// force reset the QC and DKG
+                FlowEpoch.borrowClusterQCAdmin().forceStopVoting()
+                FlowEpoch.borrowDKGAdmin().forceEndDKG()
+            }
+
+            /// Create the recovery epoch metadata, this struct is used to hold 
+            /// EpochRecover service event data.
+            let recoverEpochMetadata = FlowEpoch.RecoverEpochMetadata(
+                counter: FlowEpoch.proposedEpochCounter(),
+                nodeIDs: nodeIDs,
+                firstView: startView,
+                finalView: endView,
+                collectorClusters: collectorClusters,
+                randomSource: randomSource,
+                dkgPhase1FinalView: startView + FlowEpoch.configurableMetadata.numViewsInStakingAuction + FlowEpoch.configurableMetadata.numViewsInDKGPhase - 1 as UInt64,
+                dkgPhase2FinalView: startView + FlowEpoch.configurableMetadata.numViewsInStakingAuction + (2 as UInt64 * FlowEpoch.configurableMetadata.numViewsInDKGPhase) - 1 as UInt64,
+                dkgPhase3FinalView: startView + FlowEpoch.configurableMetadata.numViewsInStakingAuction + (3 as UInt64 * FlowEpoch.configurableMetadata.numViewsInDKGPhase) - 1 as UInt64,
+                targetDuration: targetDuration,
+                targetEndTime: targetEndTime,
+                clusterQCVoteData: clusterQCVoteData,
+                dkgPubKeys: dkgPubKeys,
+            )
```

Comment:
Remove the casts, they should not be necessary.

Maybe also extract the repeated long expressions into variables, so the code is more readable:


```suggestion
            /// Create the recovery epoch metadata, this struct is used to hold 
            /// EpochRecover service event data.
            
            let numViewsInStakingAuction = FlowEpoch.configurableMetadata.numViewsInStakingAuction
            let numViewsInDKGPhase = FlowEpoch.configurableMetadata.numViewsInDKGPhase
            
            let recoverEpochMetadata = FlowEpoch.RecoverEpochMetadata(
                counter: FlowEpoch.proposedEpochCounter(),
                nodeIDs: nodeIDs,
                firstView: startView,
                finalView: endView,
                collectorClusters: collectorClusters,
                randomSource: randomSource,
                dkgPhase1FinalView: startView + numViewsInStakingAuction + numViewsInDKGPhase - 1,
                dkgPhase2FinalView: startView + numViewsInStakingAuction + (2 * numViewsInDKGPhase) - 1,
                dkgPhase3FinalView: startView + numViewsInStakingAuction + (3 * numViewsInDKGPhase) - 1,
                targetDuration: targetDuration,
                targetEndTime: targetEndTime,
                clusterQCVoteData: clusterQCVoteData,
                dkgPubKeys: dkgPubKeys,
            )
```